### DISCUSSION TITLE
Dagger in a monorepo setup with Python build system

### DISCUSSION THREAD

**Discussion by ptrhck** [dagger/dagger#2743] [Help]
I just came accross dagger today and have a few questions. I would really appreciate some guidance whether dagger is what I am looking for, which I hope:
Currently, we are in the process of changing to a Python monorepo where we will use https://www.pantsbuild.org/ as the build system (we will have to rewrite CI/CD as well then). It is a great tool for packaging our application build artifacts from the monorepo. It also supports building docker images, but I am having the feeling that in my setup this limits us a bit. This is not due to pants but due to missing Docker features I think (mostly for DRY purposes).
I will try to give some minimum example of the setup to illustrate the issues I am facing, which I hope can be solved with dagger. So the project structure looks roughly as follows:
```
/.devcontainer
  Dockerfile
pants.toml
/django
  Dockerfile
/services
  Dockerfile
  /weasyprint
    Dockerfile
/libs
  /common
```
As you can see, we use a VSCode devcontainer on WSL2. In that container I can run something like `./pants run django` and pants will infer all Python dependencies (e.g. my common lib and third party requirements), build an artifact and run it so I can develop my application.
For deployment I will add this artifact to a Docker image (Dockerfile), define a `docker_image` target in pants and publish it to a registry. Same holds true for my service (I intentionally mentioned weasyprint, I will outline the reason below), here I will even use the the Pants Lambda integration and test this locally using localstack.
Now comes the Docker pain. Django as well as Weasyprint need some certain system dependencies. So in order to do `./pants run django `or `./pants test django` in my devcontainer those need to be added to the devcontainer Dockerfile, for the deployment also to that Dockerfile. And of course tests should run on CI as well, so they need to be available there as well. So I started drafting something like this `.devcontainer/Dockerfile` :
FROM python:3.8-bullseye as build
RUN apt-get --download-only django-sys-dependencies --folder django-build

```
FROM build as ci
# need them in ci as well as some tooling für deployment
# this target could be published to a container repository and be used as the image in CI to run all my pants and deployment commands
RUN apt-get install all .dpkg from django-build
RUN install node.js && npm install -g serverless

FROM ci as devcontainer
# just install all the stuff that has not been installed in previous stages from https://github.com/microsoft/vscode-dev-containers/blob/main/containers/python-3/.devcontainer/base.Dockerfile
```
Now in `django/Dockerfile`:

```
FROM python:3.8-bullseye
COPY --from build django-build django-build
RUN apt-get install all .dpkg
COPY artifact.pex /bin/helloworld
ENTRYPOINT ["/bin/helloworld"]
```

With pants I can define dependencies such that the build stage will be build before the Django Dockerfile will be build. Honestly, I could not think of any better solution with Docker regarding those system dependencies.
The very same thing holds true for weasyprint. So I will add the system dependencies for this service again to the first Dockerfile in the build stage and copy it to the weasyprint service Dockerfile as well. Now, I am not only having one service but multiple for which the Dockerfile in most cases just looks completely the same, where I just add that build artifact. That would be a lot of repetition.

I found this [article ](https://www.busbud.com/blog/going-docker-multi-stage-builds/) using some "hidden" feature of Docker with multi-stage + ONBUILD. I tried an noticed that this only works without Buildkit which I had enabled. Then I came accross earthly, which seems to resolve things were I reapeat a lot, in particular for Docker images. I am aware that this can be solved with [dagger ](https://docs.dagger.io/1205/container-images/#automation) in a similiar way. Am I right or did I totally missunderstand dagger? From my first impression, earthly feels way more Dockerfile focused.

As I would like to solve those Docker issues and have a CI/CD pipeline with fast iteration I would really appreciate some guidance how to use dagger in this case. Here are some questions I have:

- How does dagger fit in my devcontainer environment? We usually do not install anything in WSL itself, but install everything in the devcontainer such that this is kind of the entrypoint of development. Where do I install dagger? In the devcontainer or in that case in WSL? Is dagger then building my devcontainer? What happens when I choose "Rebuild devcontainer" in VSCode, ...
- Where do I run all those pants commands like run or test ? I have setup pre-commit hooks for linting, should they run in the devcontainer as they currently do?
- Pants can run things depending on a --changed-since=HEAD and will publish only things that truly have changed. How does that fit into an dagger setup?

**Comment by samalba** [comment]
It looks like most of your problem can be summarized as: orchestrating docker builds, which dagger can handle pretty easily. Before I dig into your points below, I'd recommend reading those 2 guides from the dagger doc:

- [Building container images](https://docs.dagger.io/1205/container-images) - basic principles to either build from Dockerfiles or directly from dagger instructions
- [The docker package](https://docs.dagger.io/1241/docker) - very detailed guide to leverage docker, including build images

I am not familiar with Pants, but I would either: integrate `pants` into a new dagger package if you want to keep using it and orchestrate some of your pants build with the rest of your dagger plan. Or re-implement what you do currently with Pants, but directly in dagger. For instance, inside the dagger plan of the official dagger repo, we use dagger directly to build binaries, lint code, run unit + integrations tests, etc... And in general, it's a good practice to rely on containers to run those, so it guarantees portability and make sure everything runs the same way everywhere (CI or different developer laptops).

> * How does dagger fit in my devcontainer environment? We usually do not install anything in WSL itself, but install everything in the devcontainer such that this is kind of the entrypoint of development. Where do I install dagger? In the devcontainer or in that case in WSL? Is dagger then building my devcontainer? What happens when I choose "Rebuild devcontainer" in VSCode, ...

I would [install dagger natively on Windows](https://docs.dagger.io/install) + use Docker Desktop. That way everything that dagger does ends up in WSL containers. Then I would build the devcontainer from dagger by integrating the vscode cli in a dagger package (let me know if you need help implementing that one).

Here is some related docs: https://code.visualstudio.com/docs/remote/devcontainer-cli#_building-a-dev-container-image - not sure how that'll play out with the "Rebuild devcontainer" button in VSCode though. Once the integration is done, you would just run `dagger do devcontainer` assuming the `devcontainer` action implements the `devcontainer build ...`.

> * Where do I run all those pants commands like run or test ? I have setup pre-commit hooks for linting, should they run in the devcontainer as they currently do?

As briefly explained above, you can run those in [dagger actions like we do on the internal CI](https://github.com/dagger/dagger/blob/main/ci.cue#L43) - the advantage is that all of them run inside container (so the behavior will be the same locally in your CI), and you can also either call all tests: `dagger do test` or just unit tests for instance: `dagger do test unit`. You can either call a group of actions or a single action by specifying its path like this.

> * Pants can run things depending on a --changed-since=HEAD and will publish only things that truly have changed. How does that fit into an dagger setup?

Dagger implements caching, so this is another thing that is simplified. You can just build container images, binaries or run tests while ignoring the new changes. Then dagger will run the actions only when the input (eg.: the source code) changed.

Let me know if you need help building some of this.

**Comment by ptrhck** [comment]
I think portability and reproducability will be a great enhancement in terms of CI/CD with dagger. I can test those pipelines locally and be sure they run on CI. And for the specific case of orchestrating docker builds for all the microservices it can be considered very helpful.

I am a little worried about two things:

1. Maintaining some wrapper around pants, i.e. through a dagger package. I have not thought about it in depth, but I would like to remain access to all pants commands
2. Deverloper experience: The mentioned advantages mostly address devops issues and running everything in dagger container seems like adding another layer to the development setup, in particular IDE integration. Adding a build system like pants already incorporates a lot more configuration, like getting the Python extension and all its features to work. I cannot think of a good way to integrate this when every command is run in an isolated container.

Appreciate your thoughts on that end!
